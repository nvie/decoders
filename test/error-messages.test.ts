import { describe, expect, test } from 'vitest';

import { formatInline, number, object, optional, positiveInteger, string } from '~';

describe('error messages (simple optional strings)', () => {
  const decoder = object({ name: optional(string) });

  test('valid', () => {
    expect(decoder.verify({})).toEqual({});
    expect(decoder.verify({ name: 'hi' })).toEqual({ name: 'hi' });
  });

  test('invalid', () => {
    expect(() => decoder.verify({ name: null })).toThrow(`
{
  "name": null,
          ^^^^
          Either:
          - Must be undefined
          - Must be string
}`);
  });
});

describe('error messages (optional objects)', () => {
  const decoder = object({
    person: optional(object({ name: string, age: positiveInteger })),
  });

  test('valid', () => {
    expect(decoder.verify({})).toEqual({});
    expect(decoder.verify({ person: { name: 'Alice', age: 29 } })).toEqual({
      person: {
        name: 'Alice',
        age: 29,
      },
    });
  });

  test('invalid', () => {
    const ann = decoder.decode({ person: { name: 'Alice', age: 'thirty' } }).error!;
    expect(formatInline(ann)).toMatchSnapshot();
  });
});
