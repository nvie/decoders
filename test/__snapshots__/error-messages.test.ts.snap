// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`error messages (optional objects) > invalid 1`] = `
"{
  "person": {
    "name": "Alice",
    "age": "thirty",
  },
            ^
            Either:
            - Must be undefined
            - Value at key 'age': Must be number
}"
`;
